File: webllm/transform.py
"""
transform.py
AST transformation ruleset for sovereign routing expressions.
Direct port of transform.mjs — same 9 rules, same priority fixpoint.
"""

import re
from dataclasses import dataclass, field
from typing import List, Optional, Tuple


@dataclass
class TransformRule:
    name: str
    priority: int

    def match(self, xml: str) -> bool:
        raise NotImplementedError

    def apply(self, xml: str) -> str:
        raise NotImplementedError


class DoubleNegationElimination(TransformRule):
    """NOT(NOT(x)) → x"""
    name = "double-negation-elimination"
    priority = 10

    def __init__(self):
        super().__init__(self.name, self.priority)
        self._pat = re.compile(r'<not>\s*<not>([\s\S]*?)</not>\s*</not>')

    def match(self, xml: str) -> bool:
        return bool(self._pat.search(xml))

    def apply(self, xml: str) -> str:
        return self._pat.sub(r'\1', xml)


class IdempotentAnd(TransformRule):
    """AND(X, X) → X"""
    name = "idempotent-and"
    priority = 9

    def __init__(self):
        super().__init__(self.name, self.priority)
        self._pat = re.compile(
            r'<and>\s*(<leaf[^>]*op="([^"]+)"[^>]*/>\s*)<leaf[^>]*op="\2"[^>]*/>\s*</and>'
        )

    def match(self, xml: str) -> bool:
        return bool(self._pat.search(xml))

    def apply(self, xml: str) -> str:
        return self._pat.sub(r'\1', xml)


class IdempotentOr(TransformRule):
    """OR(X, X) → X"""
    name = "idempotent-or"
    priority = 9

    def __init__(self):
        super().__init__(self.name, self.priority)
        self._pat = re.compile(
            r'<or>\s*(<leaf[^>]*op="([^"]+)"[^>]*/>\s*)<leaf[^>]*op="\2"[^>]*/>\s*</or>'
        )

    def match(self, xml: str) -> bool:
        return bool(self._pat.search(xml))

    def apply(self, xml: str) -> str:
        return self._pat.sub(r'\1', xml)


class OrTautology(TransformRule):
    """OR(X, NOT(X)) → X with valid=true"""
    name = "or-tautology"
    priority = 8

    def __init__(self):
        super().__init__(self.name, self.priority)
        self._pat1 = re.compile(
            r'<or>\s*(<leaf[^>]*op="([^"]+)"[^>]*/>\s*)<not>\s*<leaf[^>]*op="\2"[^>]*/>\s*</not>\s*</or>'
        )
        self._pat2 = re.compile(
            r'<or>\s*<not>\s*<leaf[^>]*op="([^"]+)"[^>]*/>\s*</not>\s*(<leaf[^>]*op="\1"[^>]*/>\s*)</or>'
        )

    def _force_valid(self, leaf: str) -> str:
        if 'valid=' in leaf:
            return re.sub(r'valid="[^"]*"', 'valid="true"', leaf)
        return leaf.replace('/>', ' valid="true"/>')

    def match(self, xml: str) -> bool:
        return bool(self._pat1.search(xml) or self._pat2.search(xml))

    def apply(self, xml: str) -> str:
        def sub1(m):
            return self._force_valid(m.group(1).strip())
        def sub2(m):
            return self._force_valid(m.group(2).strip())
        result = self._pat1.sub(sub1, xml)
        result = self._pat2.sub(sub2, result)
        return result


class AndShortCircuitFalse(TransformRule):
    """AND containing any valid=false leaf → invalid leaf"""
    name = "and-short-circuit-false"
    priority = 7
    _outer = re.compile(r'<and>([\s\S]*?)</and>')
    _false_leaf = re.compile(r'<leaf[^>]*valid="false"[^>]*/>')

    def __init__(self):
        super().__init__(self.name, self.priority)

    def match(self, xml: str) -> bool:
        return bool(re.search(r'<and>[\s\S]*?<leaf[^>]*valid="false"[^>]*/>[\s\S]*?</and>', xml))

    def apply(self, xml: str) -> str:
        def replacer(m):
            inner = m.group(1)
            if self._false_leaf.search(inner):
                return '<leaf op="ADD" weight="0.0" valid="false"/>'
            return m.group(0)
        return self._outer.sub(replacer, xml)


class OrShortCircuitTrue(TransformRule):
    """OR where first child is valid=true → emit that leaf"""
    name = "or-short-circuit-true"
    priority = 7
    _pat = re.compile(r'<or>\s*(<leaf[^>]*valid="true"[^>]*/>)[\s\S]*?</or>')

    def __init__(self):
        super().__init__(self.name, self.priority)

    def match(self, xml: str) -> bool:
        return bool(self._pat.search(xml))

    def apply(self, xml: str) -> str:
        return self._pat.sub(r'\1', xml)


class WeightNormalizationAnd(TransformRule):
    """AND where all leaves < 0.10 weight → keep highest-weight leaf"""
    name = "weight-normalization-and"
    priority = 6
    _outer = re.compile(r'<and>([\s\S]*?)</and>')
    _leaf = re.compile(r'<leaf[^>]*/>')
    _weight = re.compile(r'weight="([^"]+)"')

    def __init__(self):
        super().__init__(self.name, self.priority)

    def _all_low(self, inner: str) -> bool:
        leaves = self._leaf.findall(inner)
        if not leaves:
            return False
        return all(
            float(m.group(1)) < 0.10
            for l in leaves
            for m in [self._weight.search(l)]
            if m
        )

    def match(self, xml: str) -> bool:
        return any(self._all_low(m.group(1)) for m in self._outer.finditer(xml))

    def apply(self, xml: str) -> str:
        def replacer(m):
            inner = m.group(1)
            if not self._all_low(inner):
                return m.group(0)
            leaves = self._leaf.findall(inner)
            best = leaves[0]
            best_w = -1.0
            for l in leaves:
                wm = self._weight.search(l)
                w = float(wm.group(1)) if wm else 0.0
                if w > best_w:
                    best_w = w
                    best = l
            return best
        return self._outer.sub(replacer, xml)


class NotMemsetToXor(TransformRule):
    """NOT(MEMSET) → XOR — XOR is semantic inverse of zero-fill"""
    name = "not-memset-to-xor"
    priority = 5
    _pat = re.compile(r'<not>\s*<leaf op="MEMSET"([^>]*?)/>\s*</not>')

    def __init__(self):
        super().__init__(self.name, self.priority)

    def match(self, xml: str) -> bool:
        return bool(self._pat.search(xml))

    def apply(self, xml: str) -> str:
        return self._pat.sub(lambda m: f'<leaf op="XOR"{m.group(1)}/>', xml)


class AndMemcpyMemsetToMemcpy(TransformRule):
    """AND(MEMCPY, MEMSET) → MEMCPY — copy subsumes set"""
    name = "and-memcpy-memset-to-memcpy"
    priority = 5
    _outer = re.compile(r'<and>([\s\S]*?)</and>')
    _weight = re.compile(r'weight="([^"]+)"')

    def __init__(self):
        super().__init__(self.name, self.priority)

    def match(self, xml: str) -> bool:
        return bool(
            re.search(r'<and>[\s\S]*?<leaf op="MEMCPY"[\s\S]*?<leaf op="MEMSET"[\s\S]*?</and>', xml) or
            re.search(r'<and>[\s\S]*?<leaf op="MEMSET"[\s\S]*?<leaf op="MEMCPY"[\s\S]*?</and>', xml)
        )

    def apply(self, xml: str) -> str:
        def replacer(m):
            inner = m.group(1)
            has_copy = 'op="MEMCPY"' in inner
            has_set = 'op="MEMSET"' in inner
            if has_copy and has_set:
                wm = self._weight.search(inner)
                w = wm.group(1) if wm else '0.05'
                return f'<leaf op="MEMCPY" weight="{w}" valid="true"/>'
            return m.group(0)
        return self._outer.sub(replacer, xml)


DEFAULT_RULES = [
    DoubleNegationElimination(),
    IdempotentAnd(),
    IdempotentOr(),
    OrTautology(),
    AndShortCircuitFalse(),
    OrShortCircuitTrue(),
    WeightNormalizationAnd(),
    NotMemsetToXor(),
    AndMemcpyMemsetToMemcpy(),
]


class TransformEngine:
    def __init__(self, rules=None, max_passes: int = 10):
        self.rules = sorted(rules or DEFAULT_RULES, key=lambda r: -r.priority)
        self.max_passes = max_passes

    def transform(self, xml: str) -> dict:
        current = xml
        log = []
        for pass_n in range(self.max_passes):
            fired = False
            for rule in self.rules:
                if rule.match(current):
                    nxt = rule.apply(current)
                    if nxt != current:
                        log.append({'pass': pass_n, 'rule': rule.name})
                        current = nxt
                        fired = True
                        break
            if not fired:
                break
        return {'xml': current, 'log': log, 'passes': len(log)}

    def entropy_reduce(self, xml: str, depth_threshold: int = 3) -> dict:
        depth = self._measure_depth(xml)
        if depth <= depth_threshold:
            return {'xml': xml, 'reduced': False}
        leaves = re.findall(r'<leaf[^>]*/>', xml)
        if not leaves:
            return {'xml': xml, 'reduced': False}
        flat = '<route>\n  <or>\n    ' + '\n    '.join(leaves) + '\n  </or>\n</route>'
        return {'xml': flat, 'reduced': True, 'original_depth': depth}

    def _measure_depth(self, xml: str) -> int:
        depth = 0
        max_depth = 0
        i = 0
        while i < len(xml):
            if xml[i] != '<':
                i += 1
                continue
            if xml[i:i+2] == '</':
                depth -= 1
                end = xml.find('>', i)
                i = end + 1 if end != -1 else i + 1
            elif xml[i:i+4] in ('<!--', '<?'):
                end = xml.find('>', i)
                i = end + 1 if end != -1 else i + 1
            else:
                end = xml.find('>', i)
                if end == -1:
                    i += 1
                    continue
                tag_content = xml[i:end]
                if not tag_content.rstrip().endswith('/'):
                    depth += 1
                    max_depth = max(max_depth, depth)
                i = end + 1
        return max_depth

File: webllm/transform.mjs
// transform.mjs
// AST transformation ruleset for sovereign routing expressions
// Applies rewrite rules to <route> XML before evaluation
// Rules fire in priority order; engine runs until no rule fires (fixpoint)

import { VALID_OPS } from './ast_parser.mjs';

// Each rule: { name, match(xml), apply(xml) → xml, priority }
// match returns true if the rule applies to this XML string
// apply returns the transformed XML string

const RULES = [
  // ── priority 10 ─────────────────────────────────────────────────────────────
  {
    name: 'double-negation-elimination',
    priority: 10,
    // NOT(NOT(x)) → x
    match(xml) { return /<not>\s*<not>[\s\S]*?<\/not>\s*<\/not>/.test(xml); },
    apply(xml) {
      return xml.replace(/<not>\s*<not>([\s\S]*?)<\/not>\s*<\/not>/g, '$1');
    },
  },

  // ── priority 9 ──────────────────────────────────────────────────────────────
  {
    name: 'idempotent-and',
    priority: 9,
    // AND(x, x) → x  — two leaf children with the same op value
    match(xml) {
      return /<and>\s*<leaf[^>]*op="([^"]+)"[^>]*\/>\s*<leaf[^>]*op="\1"[^>]*\/>\s*<\/and>/.test(xml);
    },
    apply(xml) {
      // Keeps the first leaf; both are identical in op so either is correct
      return xml.replace(
        /<and>\s*(<leaf[^>]*op="([^"]+)"[^>]*\/>)\s*<leaf[^>]*op="\2"[^>]*\/>\s*<\/and>/g,
        '$1'
      );
    },
  },
  {
    name: 'idempotent-or',
    priority: 9,
    // OR(x, x) → x  — two leaf children with the same op value
    match(xml) {
      return /<or>\s*<leaf[^>]*op="([^"]+)"[^>]*\/>\s*<leaf[^>]*op="\1"[^>]*\/>\s*<\/or>/.test(xml);
    },
    apply(xml) {
      return xml.replace(
        /<or>\s*(<leaf[^>]*op="([^"]+)"[^>]*\/>)\s*<leaf[^>]*op="\2"[^>]*\/>\s*<\/or>/g,
        '$1'
      );
    },
  },

  // ── priority 8 ──────────────────────────────────────────────────────────────
  {
    name: 'or-tautology',
    priority: 8,
    // OR(x, NOT(x)) → TRUE  — emit the leaf forced to valid="true"
    // Matches both orderings: leaf-then-NOT and NOT-then-leaf
    match(xml) {
      return (
        /<or>\s*<leaf[^>]*op="([^"]+)"[^>]*\/>\s*<not>\s*<leaf[^>]*op="\1"[^>]*\/>\s*<\/not>\s*<\/or>/.test(xml) ||
        /<or>\s*<not>\s*<leaf[^>]*op="([^"]+)"[^>]*\/>\s*<\/not>\s*<leaf[^>]*op="\1"[^>]*\/>\s*<\/or>/.test(xml)
      );
    },
    apply(xml) {
      const forceValid = (leaf) =>
        leaf.includes('valid=')
          ? leaf.replace(/valid="[^"]*"/, 'valid="true"')
          : leaf.replace('/>', ' valid="true"/>');

      // Case 1: <leaf op="X".../> then <not><leaf op="X".../>...</not>
      let result = xml.replace(
        /<or>\s*(<leaf[^>]*op="([^"]+)"[^>]*\/>)\s*<not>\s*<leaf[^>]*op="\2"[^>]*\/>\s*<\/not>\s*<\/or>/g,
        (_, leaf) => forceValid(leaf)
      );
      // Case 2: <not><leaf op="X".../>...</not> then <leaf op="X".../>
      result = result.replace(
        /<or>\s*<not>\s*<leaf[^>]*op="([^"]+)"[^>]*\/>\s*<\/not>\s*(<leaf[^>]*op="\1"[^>]*\/>)\s*<\/or>/g,
        (_, _op, leaf) => forceValid(leaf)
      );
      return result;
    },
  },

  // ── priority 7 ──────────────────────────────────────────────────────────────
  {
    name: 'and-short-circuit-false',
    priority: 7,
    // AND containing any valid="false" leaf → replace whole AND with an invalid leaf
    match(xml) {
      return /<and>[\s\S]*?<leaf[^>]*valid="false"[^>]*\/>[\s\S]*?<\/and>/.test(xml);
    },
    apply(xml) {
      return xml.replace(/<and>([\s\S]*?)<\/and>/g, (match, inner) => {
        if (/<leaf[^>]*valid="false"[^>]*\/>/.test(inner)) {
          return '<leaf op="ADD" weight="0.0" valid="false"/>';
        }
        return match;
      });
    },
  },
  {
    name: 'or-short-circuit-true',
    priority: 7,
    // OR where the first direct child is a valid="true" leaf → emit that leaf
    match(xml) {
      return /<or>\s*<leaf[^>]*valid="true"[^>]*\/>/.test(xml);
    },
    apply(xml) {
      return xml.replace(
        /<or>\s*(<leaf[^>]*valid="true"[^>]*\/>)[\s\S]*?<\/or>/g,
        '$1'
      );
    },
  },

  // ── priority 6 ──────────────────────────────────────────────────────────────
  {
    name: 'weight-normalization-and',
    priority: 6,
    // All leaves in an AND have weight < 0.10 → collapse to single highest-weight leaf
    match(xml) {
      const andRe = /<and>([\s\S]*?)<\/and>/g;
      let m;
      while ((m = andRe.exec(xml)) !== null) {
        const inner = m[1];
        const leaves = [...inner.matchAll(/<leaf[^>]*\/>/g)].map(lm => lm[0]);
        if (leaves.length === 0) continue;
        const allLow = leaves.every(l => {
          const wm = l.match(/weight="([^"]+)"/);
          return wm ? parseFloat(wm[1]) < 0.10 : true;
        });
        if (allLow) return true;
      }
      return false;
    },
    apply(xml) {
      return xml.replace(/<and>([\s\S]*?)<\/and>/g, (match, inner) => {
        const leaves = [...inner.matchAll(/<leaf[^>]*\/>/g)].map(lm => lm[0]);
        if (leaves.length === 0) return match;
        const allLow = leaves.every(l => {
          const wm = l.match(/weight="([^"]+)"/);
          return wm ? parseFloat(wm[1]) < 0.10 : true;
        });
        if (!allLow) return match;
        // Find highest-weight leaf
        let bestLeaf = leaves[0];
        let bestWeight = -1;
        for (const l of leaves) {
          const wm = l.match(/weight="([^"]+)"/);
          const w = wm ? parseFloat(wm[1]) : 0;
          if (w > bestWeight) { bestWeight = w; bestLeaf = l; }
        }
        return bestLeaf;
      });
    },
  },

  // ── priority 5 ──────────────────────────────────────────────────────────────
  {
    name: 'not-memset-to-xor',
    priority: 5,
    // NOT(MEMSET) → XOR  — XOR is the semantic inverse of MEMSET (zero vs toggle)
    match(xml) {
      return /<not>\s*<leaf op="MEMSET"[^>]*\/>[\s\S]*?<\/not>/.test(xml) ||
             /<not><leaf op="MEMSET"[^>]*\/><\/not>/.test(xml);
    },
    apply(xml) {
      return xml.replace(
        /<not>\s*<leaf op="MEMSET"([^>]*?)\/>\s*<\/not>/g,
        (_, attrs) => `<leaf op="XOR"${attrs}/>`
      );
    },
  },
  {
    name: 'and-memcpy-memset-to-memcpy',
    priority: 5,
    // AND(MEMCPY, MEMSET) → MEMCPY  — copy subsumes set (either order)
    match(xml) {
      return (
        /<and>[\s\S]*?<leaf op="MEMCPY"[\s\S]*?<leaf op="MEMSET"[\s\S]*?<\/and>/.test(xml) ||
        /<and>[\s\S]*?<leaf op="MEMSET"[\s\S]*?<leaf op="MEMCPY"[\s\S]*?<\/and>/.test(xml)
      );
    },
    apply(xml) {
      return xml.replace(/<and>([\s\S]*?)<\/and>/g, (match, inner) => {
        const hasCopy = /op="MEMCPY"/.test(inner);
        const hasSet  = /op="MEMSET"/.test(inner);
        if (hasCopy && hasSet) {
          // Use the weight from the first leaf encountered
          const wMatch = inner.match(/weight="([^"]+)"/);
          const w = wMatch ? wMatch[1] : '0.05';
          return `<leaf op="MEMCPY" weight="${w}" valid="true"/>`;
        }
        return match;
      });
    },
  },
];

// ─── Transform Engine ────────────────────────────────────────────────────────

export class TransformEngine {
  constructor(rules = RULES, maxPasses = 10) {
    this.rules = [...rules].sort((a, b) => b.priority - a.priority);
    this.maxPasses = maxPasses;
  }

  /**
   * Run rules to fixpoint (no rule fires) or maxPasses, whichever comes first.
   * Returns { xml, log, passes }
   *   log   — array of { pass, rule } for every firing
   *   passes — total number of firings
   */
  transform(xml) {
    let current = xml;
    const log = [];

    for (let pass = 0; pass < this.maxPasses; pass++) {
      let fired = false;
      for (const rule of this.rules) {
        if (rule.match(current)) {
          const next = rule.apply(current);
          if (next !== current) {
            log.push({ pass, rule: rule.name });
            current = next;
            fired = true;
            break; // restart rule scan from highest priority after each firing
          }
        }
      }
      if (!fired) break;
    }

    return { xml: current, log, passes: log.length };
  }

  /**
   * Entropy reduction: if AST depth > depthThreshold, flatten all leaves
   * into a single <route><or>…</or></route> at depth 2.
   * Returns { xml, reduced, original_depth? }
   */
  entropyReduce(xml, depthThreshold = 3) {
    const depth = this._measureDepth(xml);
    if (depth <= depthThreshold) return { xml, reduced: false };

    const leaves = [...xml.matchAll(/<leaf[^>]*\/>/g)].map(m => m[0]);
    if (leaves.length === 0) return { xml, reduced: false };

    const flat = `<route>\n  <or>\n    ${leaves.join('\n    ')}\n  </or>\n</route>`;
    return { xml: flat, reduced: true, original_depth: depth };
  }

  /**
   * Measure the maximum nesting depth of the XML tree.
   * Self-closing tags do not increase depth.
   */
  _measureDepth(xml) {
    let depth = 0;
    let max = 0;
    let i = 0;

    while (i < xml.length) {
      if (xml[i] !== '<') { i++; continue; }

      const next = xml[i + 1];

      if (next === '/') {
        // Closing tag — decrease depth
        depth--;
        const end = xml.indexOf('>', i);
        i = end !== -1 ? end + 1 : i + 1;
      } else if (next === '!' || next === '?') {
        // Comment / processing instruction — skip
        const end = xml.indexOf('>', i);
        i = end !== -1 ? end + 1 : i + 1;
      } else {
        // Opening or self-closing tag
        const end = xml.indexOf('>', i);
        if (end === -1) { i++; continue; }

        // Self-closing if the character before '>' (ignoring whitespace) is '/'
        const tagContent = xml.slice(i, end); // excludes the '>'
        const isSelfClosing = tagContent.trimEnd().endsWith('/');

        if (!isSelfClosing) {
          depth++;
          if (depth > max) max = depth;
        }
        i = end + 1;
      }
    }

    return max;
  }
}

export { RULES };
